File: c/src/grains.c
#include "grains.h"
#include <math.h>
#include <stdint.h>

uint64_t square(uint8_t index) {
    return (index > 0 && index < 65) ? pow(2.0, index - 1) : 0;
}

uint64_t total() {
    uint64_t two63 = pow(2.0, 63);
    return two63 - 1ull + two63;
}

File: scala/src/main/scala/Grains.scala
object Grains {
  def square(square:Int):Option[BigInt] = {
    if (square > 0 && square <= 64) {
      Some(BigInt(1) << (square - 1))
    } else {
      None
    }
  }

  def total:BigInt = {
    (BigInt(1) << 64) - 1
  }
}